CB-13685 android: Adaptive Icon Support

erisu opened the pull request:
<!--
Please make sure the checklist boxes are all checked before submitting the PR. The checklist
is intended as a quick reference, for complete details please see our Contributor Guidelines:

http://cordova.apache.org/contribute/contribute_guidelines.html

Thanks!
-->

### Platforms affected
Android

### What does this PR do?
This PR adds support for Adaptive Icon.

Changes to Default Template Project
- Updated AndroidManifest to use adaptive icons configuration by default.
- Added adaptive icons and ic_launcher.xml for Android platform 26 and higher.

Added JavaScript Unit Testing.

**config.xml for Adaptive Icons with Images**
```
<platform name="android">
  <icon background="res/icon/android/ldpi-background.png" density="ldpi" foreground="res/icon/android/ldpi-foreground.png" />
  <icon background="res/icon/android/mdpi-background.png" density="mdpi" foreground="res/icon/android/mdpi-foreground.png" />
  <icon background="res/icon/android/hdpi-background.png" density="hdpi" foreground="res/icon/android/hdpi-foreground.png" />
  <icon background="res/icon/android/xhdpi-background.png" density="xhdpi" foreground="res/icon/android/xhdpi-foreground.png" />
  <icon background="res/icon/android/xxhdpi-background.png" density="xxhdpi" foreground="res/icon/android/xxhdpi-foreground.png" />
  <icon background="res/icon/android/xxxhdpi-background.png" density="xxxhdpi" foreground="res/icon/android/xxxhdpi-foreground.png" />
</platform>
```

In this case, the foreground image will automatically be applied to the src attribute value. This is to support non-adaptive icon supported Android devices. If the user supplies their own src value, foreground will not be copied over.

**config.xml for Adaptive Icons with Vectors**
```
<platform name="android">
  <icon background="res/icon/android/ldpi-background.xml” density="ldpi" foreground="res/icon/android/ldpi-foreground.xml” src="res/icon/android/ldpi-icon.png" />
  <icon background="res/icon/android/mdpi-background.xml” density="mdpi" foreground="res/icon/android/mdpi-foreground.xml” src="res/icon/android/mdpi-icon.png" />
  <icon background="res/icon/android/hdpi-background.xml” density="hdpi" foreground="res/icon/android/hdpi-foreground.xml” src="res/icon/android/hdpi-icon.png" />
  <icon background="res/icon/android/xhdpi-background.xml” density="xhdpi" foreground="res/icon/android/xhdpi-foreground.xml” src="res/icon/android/xdpi-icon.png" />
  <icon background="res/icon/android/xxhdpi-background.xml” density="xxhdpi" foreground="res/icon/android/xxhdpi-foreground.xml” src="res/icon/android/xxhdpi-icon.png" />
  <icon background="res/icon/android/xxxhdpi-background.xml” density="xxxhdpi" foreground="res/icon/android/xxxhdpi-foreground.xml” src="res/icon/android/xxxhdpi-icon.png" />
</platform>
```

In this example, the adaptive icon supports vectors and colors for the foreground and background. If the foreground is a vector or color, the src attribute must also be defined with a PNG for Android devices that does not support adaptive icon.

**config.xml for Adaptive Icons with Colors for Background**
First: create colors.xml file with content
```
<?xml version="1.0" encoding="utf-8"?>
<resources>
    <color name="background">#FF0000</color>
</resources>
```
Second: update config.xml with…
```
<platform name="android">
  <resource-file src="/res/values/colors.xml" target="/app/src/main/res/values/colors.xml" />

  <icon background=“@color/background” density="ldpi" foreground="res/icon/android/ldpi-foreground.png" />
  <icon background="@color/background" density="mdpi" foreground="res/icon/android/mdpi-foreground.png" />
  <icon background="@color/background" density="hdpi" foreground="res/icon/android/hdpi-foreground.png" />
  <icon background="@color/background" density="xhdpi" foreground="res/icon/android/xhdpi-foreground.png" />
  <icon background="@color/background" density="xxhdpi" foreground="res/icon/android/xxhdpi-foreground.png" />
  <icon background="@color/background" density="xxxhdpi" foreground="res/icon/android/xxxhdpi-foreground.png" />
</platform>
```

The naming convention for the two new attributes, `background` and `foreground`, was decided to match the element names that is used in the `ic_launcher.xml` for consistency. https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive#creating_adaptive_icons_in_xml

** Error Example: Foreground contains vector and missing src attribute.**

```
<platform name="android">
  <icon background="res/icon/android/mdpi-background.png" foreground="res/icon/android/mdpi-foreground.xml” density=“mdpi" />
</platform>
```

### What testing has been done on this change?
- Built default app with no modifications to config.xml
- Built default app with legacy icon set defined
- Built default app with adaptive icon set with images.
- Built default app with adaptive icon set with vectors.
- Built default app with adaptive icon set with colors.
- Converted from adaptive to legacy icon set and rebuilt.
- Converted from legacy to adaptive icon set and rebuild.


### Checklist
- [X] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [X] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [X] Added automated test coverage as appropriate for this change.

erisu commented:
I also want to point out that this PR will be required. 

https://github.com/apache/cordova-common/pull/26

This PR contains the updates to the ConfigParser to include `foreground` and `background` when getting static resources.

dpogue commented:
You can ignore the node 4 Appveyor failure due to `let`/`const`, we're in the process of dropping support for node 4.

erisu commented:
@dpogue I have also noticed this.

There is that CI improvement PR, https://github.com/apache/cordova-android/pull/442, which also deprecates node 4 and adds node 10 support, among other changes.

When I testing my PR with the CI improvement PR, I saw other issues also.
https://ci.appveyor.com/project/erisu/cordova-android/build/1.0.14

If we just remove node 4 and add node 10 with no additional changes, then my PR passed.
https://ci.appveyor.com/project/erisu/cordova-android/build/1.0.17

raphinesse commented:
@erisu I rebased #442 onto latest master and your changes on top. CI tests pass on [Travis](https://travis-ci.org/raphinesse/cordova-android/builds/391468027) as well as [AppVeyor](https://ci.appveyor.com/project/raphinesse/cordova-android/build/1.0.71). I guess the failures were fixed by #450.

erisu commented:
@raphinesse I rebased with the latest master now that the #442 PR has been merged. Do you think this is ready to merge?

raphinesse commented:
Sorry, I know very little about the Android build, so I can't really review this. Maybe @dpogue or @infiloop.

raphinesse commented on spec/unit/prepare.spec.js:
```diff
@@ -0,0 +1,935 @@
+                height: undefined,
+                background: 'res/icon/android/mdpi-background.png',
+                foreground: undefined
+            }];
```
You have something similar to this in almost every test. Might be nice to factor this out e.g as a function that fills in missing props as undefined. This could help to slim down this 1000 lines spec.

raphinesse commented on spec/unit/prepare.spec.js:
```diff
@@ -0,0 +1,935 @@
+            new CordovaError('One of the following attributes are set but missing the other for the density type: mdpi. Please ensure that all require attributes are defined.')
+        );
+
+        done();
```
Are these tests asynchronous? If not, we don't need this. If so, I suggest returning promises to Jasmine instead of using the done callback.

erisu commented:
@raphinesse Thank you for pointing out improvements to my test spec. I have updated the test and reduced the line count. I also removed the `done()` calls as you pointed out because they were not needed.

Menardi commented:
This PR looks good to me. I need adaptive icons for one of my own apps, so I'm eager to see this get merged soon. Without support for this, icons for apps using `cordova-android@7` (targeting API 26+) look a bit rubbish on newer devices.

I noticed the main cordova-android maintainer, Joe, has announced on the mailing list that he is stepping down. So, I expect that getting this thoroughly reviewed is now more difficult. If there is anything I can do to get this over the line, please let me know.

dpogue commented:
I will make sure this gets merged into the next major

janpio commented:
@raphinesse Could you have a look again and redo your review please?

janpio commented:
Thanks @raphinesse! Same for me... hope someone with actual Android experience turns up.

erisu commented:
@janpio I rebased with the latest master.  I also did the same for `cordova-common`'s PR that is associated with this PR. 

The downside, both `cordova-common` and `cordova-android` have `jasmine@^3.1.0` as the dep. Both will update to `3.2.0`, and It seems the `cordova-common` tests already failed for what I believe is the same side-effect of the latest Jasmine's release.

https://github.com/apache/cordova-common/pull/26

I am expecting these tests might have an issue too. We might need to update both repo's with the temp fix...

dpogue commented:
The changes in this PR look fine to me. I had held off on merging due to some of the questions on the cordova-common side about where the Android-specific handling code should live, especially with @erisu's proposed refactoring.

janpio commented:
Are you using cordova-android `master` in your app? This hasn't really been released yet.

wisespira commented:
Oops thanks I'll again with the master branch

erisu commented:
Though the PR is merged into master, the `package.json` still references `cordova-common@^2.2.0`. The `cordova-common` PR that is necessary for this change is also merged into master which will become `cordova-common@3.0.0`.

Using `cordova-android@master` will not work on its own unless you link `cordova-common@master`. Or wait for release.

LarryBJohnson commented:
How would one go about switching from slandered npm cordova to this master branch?

janpio commented:
What does `slandered npm cordova` mean?

LarryBJohnson commented:
If I where to go in my command line and download cordova using node js I would type npm install -g cordova would this give me a version of cordova that would let me use adaptive icons ?

janpio commented:
No.

raphinesse commented:
No. `npm i -g cordova@nightly` and `npm i cordova-android@nightly` in your project

LarryBJohnson commented:
so does adding @nightly switch it to this branch ?

raphinesse commented:
No. It uses the latest nightly build. Should be easier to setup than using master. Try an earlier nightly build if npm throws errors at you during installation. A full list of available versions can be found via npm.

nprail commented:
Has this been released yet? If not, when will it be?

dpogue commented:
It is not released yet, it will be included in the next major version. We can't/won't commit to any particular timeline for that next major, but we are working on it.
